CMake Error: No source directory specified for -S  (VSC-1004)

### OS

Windows

### Operating System version

Windows 10

### Visual Studio Code version

1.73.0

### ESP-IDF version

latest, master branch

### Python version

3.9

### Doctor command output

---------------------------------------------- ESP-IDF Extension for Visual Studio Code report ---------------------------------------------
OS win32 x64 10.0.19043 
System environment variable IDF_PYTHON_ENV_PATH 
 undefined 
System environment variable PATH 
 D:\Python39\Scripts\;D:\Python39\;%ALLEGRO_BRD2ODB%\nv\bin;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\windows\System32\OpenSSH\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;c:\Program Files (x86)\HP\HP Performance Advisor;D:\PuTTY\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;C:\Program Files (x86)\IVI Foundation\VISA\WinNT\Bin\;C:\Program Files\IVI Foundation\VISA\Win64\Bin\;C:\Program Files (x86)\IVI Foundation\VISA\WinNT\Bin;D:\Git\cmd;C:\Program Files (x86)\STMicroelectronics\STM32 ST-LINK Utility\ST-LINK Utility;C:\Program Files (x86)\IVI Foundation\IVI\bin;C:\Program Files\IVI Foundation\IVI\bin;D:\CMake\bin;D:\doxygen\bin;D:\Mscgen;D:\stlink-1.3.0-win64\bin;C:\Program Files (x86)\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\Intel\Intel(R) Management Engine Components\DAL;C:\ProgramData\chocolatey\bin;D:\nodejs\;C:\Program Files\Docker\Docker\resources\bin;C:\ProgramData\DockerDesktop\version-bin;C:\Program Files\usbipd-win\;C:\Program Files (x86)\Nordic Semiconductor\nrf-command-line-tools\bin\;D:\Keil_v5\ARM\ARMCC\bin;D:\serialplot\bin;C:<HOMEPATH>\AppData\Local\Microsoft\WindowsApps;D:\Microsoft VS Code\bin;C:<HOMEPATH>\AppData\Roaming\npm 
System environment variable PYTHON 
 undefined 
Visual Studio Code version 1.73.0 
Visual Studio Code language en 
Visual Studio Code shell C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe 
ESP-IDF Extension version 1.5.1 
Workspace folder f:\BLE_WIFI\Espressif\SDK\advertising\non-connectable-scannale-undirected-adv 
---------------------------------------------------- Extension configuration settings ------------------------------------------------------
ESP-ADF Path (idf.espAdfPath) ${env:ADF_PATH}
ESP-IDF Path (idf.espIdfPath) f:\BLE_WIFI\Espressif\SDK\esp-idf
ESP-MDF Path (idf.espMdfPath) ${env:MDF_PATH}
ESP-Matter Path (idf.espMatterPath) ${env:ESP_MATTER_PATH}
Custom extra paths (idf.customExtraPaths) F:\BLE_WIFI\Espressif\SDK\esp-idf\components\esptool_py\esptool;F:\BLE_WIFI\Espressif\SDK\esp-idf\components\app_update;F:\BLE_WIFI\Espressif\SDK\esp-idf\components\espcoredump;F:\BLE_WIFI\Espressif\SDK\esp-idf\components\partition_table;F:\Tools\Espressif\IDF_Tools\.espressif\tools\riscv32-esp-elf-gdb\11.2_20220823\riscv32-esp-elf-gdb\bin;F:\Tools\Espressif\IDF_Tools\.espressif\tools\riscv32-esp-elf\esp-2022r1-11.2.0\riscv32-esp-elf\bin;F:\Tools\Espressif\IDF_Tools\.espressif\tools\cmake\3.24.0\bin;F:\Tools\Espressif\IDF_Tools\.espressif\tools\openocd-esp32\v0.11.0-esp32-20220706\openocd-esp32\bin;F:\Tools\Espressif\IDF_Tools\.espressif\tools\ninja\1.10.2\;F:\Tools\Espressif\IDF_Tools\.espressif\tools\idf-exe\1.0.3\;F:\Tools\Espressif\IDF_Tools\.espressif\tools\ccache\4.6.2\ccache-4.6.2-windows-x86_64;F:\Tools\Espressif\IDF_Tools\.espressif\python_env\idf5.1_py3.9_env\Scripts;F:\BLE_WIFI\Espressif\SDK\esp-idf\tools;
Custom extra vars (idf.customExtraVars) {"OPENOCD_SCRIPTS":"F:/Tools/Espressif/IDF_Tools/.espressif/tools/openocd-esp32/v0.11.0-esp32-20220706/openocd-esp32/share/openocd/scripts","IDF_CCACHE_ENABLE":"1"}
Virtual env Python Path (idf.pythonBinPath) F:\Tools\Espressif\IDF_Tools\.espressif\python_env\idf5.1_py3.9_env\Scripts\python.exe
Serial port (idf.port) COM35
OpenOCD Configs (idf.openOcdConfigs) board/esp32c3-builtin.cfg
ESP-IDF Tools Path (idf.toolsPath) f:\Tools\Espressif\IDF_Tools\.espressif
Git Path (idf.gitPath) ${env:programfiles}\Git\cmd\git.exe
-------------------------------------------------------- Configurations access -------------------------------------------------------------
Access to ESP-ADF Path (idf.espAdfPath) false
Access to ESP-IDF Path (idf.espIdfPath) true
Access to ESP-MDF Path (idf.espMdfPath) false
Access to ESP-Matter Path (idf.espMatterPath) false
Access to ESP-IDF Custom extra paths
Access to F:\BLE_WIFI\Espressif\SDK\esp-idf\components\esptool_py\esptool: true
Access to F:\BLE_WIFI\Espressif\SDK\esp-idf\components\app_update: true
Access to F:\BLE_WIFI\Espressif\SDK\esp-idf\components\espcoredump: true
Access to F:\BLE_WIFI\Espressif\SDK\esp-idf\components\partition_table: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\tools\riscv32-esp-elf-gdb\11.2_20220823\riscv32-esp-elf-gdb\bin: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\tools\riscv32-esp-elf\esp-2022r1-11.2.0\riscv32-esp-elf\bin: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\tools\cmake\3.24.0\bin: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\tools\openocd-esp32\v0.11.0-esp32-20220706\openocd-esp32\bin: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\tools\ninja\1.10.2\: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\tools\idf-exe\1.0.3\: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\tools\ccache\4.6.2\ccache-4.6.2-windows-x86_64: true
Access to F:\Tools\Espressif\IDF_Tools\.espressif\python_env\idf5.1_py3.9_env\Scripts: true
Access to F:\BLE_WIFI\Espressif\SDK\esp-idf\tools: true
Access to : false
Access to Virtual env Python Path (idf.pythonBinPath) true
Access to CMake in environment PATH undefined
Access to Ninja in environment PATH undefined
Access to ESP-IDF Tools Path (idf.toolsPath) true
-------------------------------------------------------- Configurations has spaces -------------------------------------------------------------
Spaces in system environment Path true
Spaces in ESP-ADF Path (idf.espAdfPath) false
Spaces in ESP-IDF Path (idf.espIdfPath) false
Spaces in ESP-MDF Path (idf.espMdfPath) false
Spaces in ESP-Matter Path (idf.espMatterPath) false
Spaces in ESP-IDF Custom extra paths
Spaces in F:\BLE_WIFI\Espressif\SDK\esp-idf\components\esptool_py\esptool: false
Spaces in F:\BLE_WIFI\Espressif\SDK\esp-idf\components\app_update: false
Spaces in F:\BLE_WIFI\Espressif\SDK\esp-idf\components\espcoredump: false
Spaces in F:\BLE_WIFI\Espressif\SDK\esp-idf\components\partition_table: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\tools\riscv32-esp-elf-gdb\11.2_20220823\riscv32-esp-elf-gdb\bin: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\tools\riscv32-esp-elf\esp-2022r1-11.2.0\riscv32-esp-elf\bin: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\tools\cmake\3.24.0\bin: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\tools\openocd-esp32\v0.11.0-esp32-20220706\openocd-esp32\bin: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\tools\ninja\1.10.2\: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\tools\idf-exe\1.0.3\: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\tools\ccache\4.6.2\ccache-4.6.2-windows-x86_64: false
Spaces in F:\Tools\Espressif\IDF_Tools\.espressif\python_env\idf5.1_py3.9_env\Scripts: false
Spaces in F:\BLE_WIFI\Espressif\SDK\esp-idf\tools: false
Spaces in : false
Spaces in Virtual env Python Path (idf.pythonBinPath) false
Spaces in ESP-IDF Tools Path (idf.toolsPath) false
----------------------------------------------------------- Executables Versions -----------------------------------------------------------
Git version undefined
ESP-IDF version undefined
Python version undefined
Python's pip version undefined
-------------------------------------------------- Python packages in idf.pythonBinPath ----------------------------------------------------
---------------------------------------------------- Check ESP-IDF python requirements.txt -------------------------------------------------
Check ESP-IDF Python packages undefined
---------------------------------------------------- Check extension requirements.txt ------------------------------------------------------
Check Extension Python packages undefined
---------------------------------------------------- Check ESP-IDF debug adapter requirements.txt ------------------------------------------
Check Debug AdapterPython packages undefined
----------------------------------------------------------- Latest error -----------------------------------------------------------------
Latest error at Unknown error in ESP-IDF doctor command
--------------------------------------------------------------------------------------------------------------------------------------------


### Extension

NULL

### Description

I upgraded the extension from v1.5.0 to v1.5.1 yesterday, however it cannot build now and the error info as below:
```
CMake Error: No source directory specified for -S
CMake Error: Run 'cmake --help' for all supported options.
```
However, it works well after rolling back to v1.5.0. could you help me dobule-check what happened?

### Debug Message

```plain
NULL
```


### Other Steps to Reproduce

NULL

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

I have the same issue:
```
Task wird im Ordner node_module_project ausgeführt: cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B /home/ih303/Projekte/esp-module/node_module_project/build -S -DCCACHE_ENABLE=1 /home/ih303/Projekte/esp-module/node_module_project 

CMake Error: No source directory specified for -S
CMake Error: Run 'cmake --help' for all supported options.
```

**-DCCACHE_ENABLE=1** is inserted between **-S** and the source directory


I think @xiaolongba is right:
> it works well after that, so I think you need to modify the order of Cmake options parameter.

--- a/src/build/buildTask.ts
+++ b/src/build/buildTask.ts
@@ -149,12 +149,13 @@ export class BuildTask {
         "-DPYTHON_DEPS_CHECKED=1",
         "-DESP_PLATFORM=1",
       ];
-      compilerArgs.push(
-        "-B",
-        this.buildDirPath,
-        "-S",
-        this.curWorkspace.fsPath
-      );
+      if (compilerArgs.indexOf("-B") === -1) {
+        compilerArgs.push("-B", this.buildDirPath);
+      }
+
+      if (compilerArgs.indexOf("-S") === -1) {
+        compilerArgs.push("-S", this.curWorkspace.fsPath);
+      }
       const enableCCache = idfConf.readParameter(
         "idf.enableCCache",
         this.curWorkspace
@@ -162,7 +163,7 @@ export class BuildTask {
       if (enableCCache && compilerArgs && compilerArgs.length) {
         const indexOfCCache = compilerArgs.indexOf("-DCCACHE_ENABLE=1");
         if (indexOfCCache === -1) {
-          compilerArgs.splice(compilerArgs.length - 1, 0, "-DCCACHE_ENABLE=1");
+          compilerArgs.push("-DCCACHE_ENABLE=1");
         }
       }
       const compileExecution = this.getShellExecution(compilerArgs, options);

